### PR TITLE
Enable modification of sysResourceRole

### PR DESCRIPTION
**Pull request by msporny**
Please check to make sure adding capabilities to sysResourceRole is implemented properly.

### CODE REVIEW

**Comment by dlongley** [comment]
I've made changes here, please review: @msporny, @davidlehn, @mattcollier

**Comment by dlongley** [comment]
The one change I could potentially think of here in terms of future proofing would be to try and generalize the `identity.delegateCapability.translate` event.

File: lib/config.js [hunk 1]
```diff
@@ -77,3 +79,8 @@ permissions.IDENTITY_REMOVE = {
   label: 'Remove Identity',
   comment: 'Required to remove an Identity.'
 };
+permissions.IDENTITY_CAPABILITY_DELEGATE = {
```
Comment (msporny) [inline lib/config.js:82]: Did this replace IDENTITY_ROLE_DELEGATE?

File: lib/index.js [hunk 2]
```diff
@@ -483,18 +512,193 @@ api.checkPermission = function(actor, permission, options, callback) {
...
+      for(const operation of options.changes) {
+        if(operation.op === 'set') {
+          bedrock.util.extend(identity, operation.value);
+          continue;
+        }
+        if(operation.op === 'add') {
+          for(const key in operation.value) {
+            const field = `identity.${key}`;
+            if(field in pushFields) {
+              pushFields[field].$each.push(...operation.value[key]);
```
Comment (msporny) [inline lib/index.js:553]: Is this witchcraft?!

File: lib/index.js [hunk 3]
```diff
@@ -523,6 +727,133 @@ function _ensureMembershipValid(actor, identity, callback) {
...
+    identity.sysResourceRole = [identity.sysResourceRole];
+  }
+  if(actor === null) {
+    // optimization to skip all permission checks; `actor` of null has
+    // full administrative permission
+    return callback();
+  }
+  async.auto({
+    cacheRoles: callback => {
+      if(options.cache.roles) {
```
Comment (msporny) [inline lib/index.js:760]: Does this create problems for role admin tools?

File: test/mocha/10-api.js [hunk 4]
```diff
@@ -607,6 +637,82 @@ describe('bedrock-identity', function() {
...
+      it('should allow identity w/ sysResourceRole for a group', done => {
+        var userName = '814b3e02-db14-4fa2-b45b-5d8f977ac087';
+        var groupName = '8314e48c-ff51-427e-aa7b-524cda45b708';
+        var newIdentity = helpers.createIdentity(userName);
+        var newGroup = helpers.createIdentity(groupName);
+        newGroup.type = ['Identity', 'Group'];
+        newGroup.owner = newIdentity.id;
+        newIdentity.memberOf = [newGroup.id];
+        newIdentity.sysResourceRole.push({
+          sysRole: 'bedrock-identity.regular',
```
Comment (msporny) [inline test/mocha/10-api.js:686]: Where did you end up wrt. your thinking here? Are we going to eventually collapse these two into a single sysResourceRole entry, or always keep them separate?

File: test/mocha/10-api.js [hunk 5]
```diff
@@ -726,6 +877,145 @@ describe('bedrock-identity', function() {
...
+              err.name.should.equal('PermissionDenied');
+              err.details.sysPermission.should.equal(
+                'IDENTITY_UPDATE_MEMBERSHIP');
+              err.details.should.be.an('object');
+              callback();
+            });
+          }]
+        }, done);
+      });
+      it('should allow identity update w/ sysResourceRole for owned group',
```
Comment (msporny) [inline test/mocha/10-api.js:1032]: Do we have tests for update of sysResourceRole via legacy API vs. changeset API?

File: lib/config.js (hunk 1 above)
Comment (dlongley) [reply lib/config.js:82]: Yes

File: lib/index.js (hunk 2 above)
Comment (dlongley) [reply lib/index.js:553]: Yes

File: lib/index.js (hunk 3 above)
Comment (dlongley) [reply lib/index.js:760]: No, it's used during a single call, it's not more persistent than that

File: test/mocha/10-api.js (hunk 4 above)
Comment (dlongley) [reply test/mocha/10-api.js:686]: Only collapse when it's the same `delegator` (I record who did the delegating now to allow for revocation in the future) ... we don't have any code that does that collapsing or uniqueness checking yet, but that's what we'd do in the future.

File: test/mocha/10-api.js (hunk 5 above)
Comment (dlongley) [reply test/mocha/10-api.js:1032]: No, there's no way to do it with the legacy API, and we should really remove the membership tests too, but I didn't want to break anything that had been out there for a bit. I assume the `sysResourceRole` stuff was new enough to just rip out and not support any legacy things.

**Pull request merged**